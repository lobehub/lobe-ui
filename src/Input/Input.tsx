'use client';

import { Input as AntInput } from 'antd';
import { cva } from 'class-variance-authority';
import { memo, useMemo } from 'react';

import { useStyles } from './style';
import type { InputProps } from './type';

const Input = memo<InputProps>(({ ref, variant = 'outlined', shadow, className, ...rest }) => {
  const { styles, cx } = useStyles();

  const variants = useMemo(
    () =>
      cva(styles.root, {
        defaultVariants: {
          shadow: false,
          variant: 'filled',
        },
        /* eslint-disable sort-keys-fix/sort-keys-fix */
        variants: {
          variant: {
            filled: styles.filled,
            outlined: styles.outlined,
            borderless: styles.borderless,
            underlined: null,
          },
          shadow: {
            false: null,
            true: styles.shadow,
          },
        },
        /* eslint-enable sort-keys-fix/sort-keys-fix */
      }),
    [styles],
  );

  return (
    <AntInput
      className={cx(variants({ shadow, variant }), className)}
      ref={ref}
      variant={variant}
      {...rest}
    />
  );
});

Input.displayName = 'Input';

export default Input;
