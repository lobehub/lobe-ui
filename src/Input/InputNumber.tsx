'use client';

import { InputNumber as AntInputNumber } from 'antd';
import { cva } from 'class-variance-authority';
import { memo, useMemo } from 'react';

import { useStyles } from './style';
import type { InputNumberProps } from './type';

const InputNumber = memo<InputNumberProps>(({ ref, variant, shadow, className, ...rest }) => {
  const { styles, cx, theme } = useStyles();

  const variants = useMemo(
    () =>
      cva(styles.root, {
        defaultVariants: {
          shadow: false,
        },
        /* eslint-disable sort-keys-fix/sort-keys-fix */
        variants: {
          variant: {
            filled: styles.filled,
            outlined: styles.outlined,
            borderless: styles.borderless,
            underlined: null,
          },
          shadow: {
            false: null,
            true: styles.shadow,
          },
        },
        /* eslint-enable sort-keys-fix/sort-keys-fix */
      }),
    [styles],
  );

  return (
    <AntInputNumber
      className={cx(
        variants({ shadow, variant: variant || (theme.isDarkMode ? 'filled' : 'outlined') }),
        className,
      )}
      ref={ref}
      variant={variant || (theme.isDarkMode ? 'filled' : 'outlined')}
      {...rest}
    />
  );
});

InputNumber.displayName = 'InputNumber';

export default InputNumber;
