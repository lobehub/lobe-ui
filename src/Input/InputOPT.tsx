'use client';

import { Input as AntInput } from 'antd';
import { cva } from 'class-variance-authority';
import { memo, useMemo } from 'react';

import { useStyles } from './style';
import type { InputOPTProps } from './type';

const InputOPT = memo<InputOPTProps>(({ ref, variant, shadow, className, ...rest }) => {
  const { styles, cx, theme } = useStyles();

  const variants = useMemo(
    () =>
      cva(styles.rootOPT, {
        defaultVariants: {
          shadow: false,
        },
        /* eslint-disable sort-keys-fix/sort-keys-fix */
        variants: {
          variant: {
            filled: styles.filledOPT,
            outlined: styles.outlinedOPT,
            borderless: styles.borderlessOPT,
            underlined: null,
          },
          shadow: {
            false: null,
            true: styles.shadowOPT,
          },
        },
        /* eslint-enable sort-keys-fix/sort-keys-fix */
      }),
    [styles],
  );

  return (
    <AntInput.OTP
      className={cx(
        variants({ shadow, variant: variant || (theme.isDarkMode ? 'filled' : 'outlined') }),
        className,
      )}
      ref={ref}
      variant={variant || (theme.isDarkMode ? 'filled' : 'outlined')}
      {...rest}
    />
  );
});

InputOPT.displayName = 'InputOPT';

export default InputOPT;
