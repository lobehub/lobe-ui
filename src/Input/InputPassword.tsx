'use client';

import { Input as AntInput } from 'antd';
import { cva } from 'class-variance-authority';
import { memo, useMemo } from 'react';

import { useStyles } from './style';
import type { InputPasswordProps } from './type';

const InputPassword = memo<InputPasswordProps>(({ ref, variant, shadow, className, ...rest }) => {
  const { styles, cx, theme } = useStyles();

  const variants = useMemo(
    () =>
      cva(styles.root, {
        defaultVariants: {
          shadow: false,
          variant: 'filled',
        },
        /* eslint-disable sort-keys-fix/sort-keys-fix */
        variants: {
          variant: {
            filled: styles.filled,
            outlined: styles.outlined,
            borderless: styles.borderless,
            underlined: null,
          },
          shadow: {
            false: null,
            true: styles.shadow,
          },
        },
        /* eslint-enable sort-keys-fix/sort-keys-fix */
      }),
    [styles],
  );

  return (
    <AntInput.Password
      className={cx(
        variants({ shadow, variant: variant || (theme.isDarkMode ? 'filled' : 'outlined') }),
        className,
      )}
      ref={ref}
      variant={variant}
      {...rest}
    />
  );
});

InputPassword.displayName = 'InputPassword';

export default InputPassword;
