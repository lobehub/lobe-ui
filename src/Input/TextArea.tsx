'use client';

import { Input as AntInput } from 'antd';
import { cva } from 'class-variance-authority';
import { memo, useMemo } from 'react';

import { useStyles } from './style';
import type { TextAreaProps } from './type';

const TextArea = memo<TextAreaProps>(
  ({ ref, variant, shadow, className, resize = false, style, ...rest }) => {
    const { styles, cx, theme } = useStyles();

    const variants = useMemo(
      () =>
        cva(styles.root, {
          defaultVariants: {
            shadow: false,
          },
          /* eslint-disable sort-keys-fix/sort-keys-fix */
          variants: {
            variant: {
              filled: styles.filled,
              outlined: styles.outlined,
              borderless: styles.borderless,
              underlined: null,
            },
            shadow: {
              false: null,
              true: styles.shadow,
            },
          },
          /* eslint-enable sort-keys-fix/sort-keys-fix */
        }),
      [styles],
    );

    return (
      <AntInput.TextArea
        className={cx(
          variants({ shadow, variant: variant || (theme.isDarkMode ? 'filled' : 'outlined') }),
          className,
        )}
        ref={ref}
        style={{ resize: resize ? undefined : 'none', ...style }}
        variant={variant || (theme.isDarkMode ? 'filled' : 'outlined')}
        {...rest}
      />
    );
  },
);

TextArea.displayName = 'TextArea';

export default TextArea;
