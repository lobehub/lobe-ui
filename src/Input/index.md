---
nav: Components
group: Data Entry
title: Input
description: A basic widget for getting the user input is a text field. Keyboard and mouse can be used for providing or changing data.
---

## Default

The rest of the props of Input are exactly the same as the original [input](https://ant.design/components/input).

<code src="./demos/index.tsx" nopadding></code>

## TextArea

<code src="./demos/TextArea.tsx" nopadding></code>

## InputNumber

<code src="./demos/InputNumber.tsx" nopadding></code>

## InputPassword

<code src="./demos/InputPassword.tsx" nopadding></code>

## InputOPT

<code src="./demos/InputOPT.tsx" nopadding></code>

## APIs

### Input

| Property | Description                        | Type                                     | Default                                            |
| -------- | ---------------------------------- | ---------------------------------------- | -------------------------------------------------- |
| shadow   | Enable shadow effect for the input | `boolean`                                | `false`                                            |
| variant  | Style variant of the input         | `'filled' \| 'outlined' \| 'borderless'` | Dark mode: `'filled'`<br/>Light mode: `'outlined'` |

Input inherits all properties from Ant Design's Input component.

### Input.TextArea

| Property | Description                           | Type                                     | Default                                            |
| -------- | ------------------------------------- | ---------------------------------------- | -------------------------------------------------- |
| shadow   | Enable shadow effect for the textarea | `boolean`                                | `false`                                            |
| variant  | Style variant of the textarea         | `'filled' \| 'outlined' \| 'borderless'` | Dark mode: `'filled'`<br/>Light mode: `'outlined'` |
| resize   | Whether textarea is resizable         | `boolean`                                | `false`                                            |

TextArea inherits all properties from Ant Design's Input.TextArea component.

### Input.Number

| Property | Description                        | Type                                     | Default                                            |
| -------- | ---------------------------------- | ---------------------------------------- | -------------------------------------------------- |
| shadow   | Enable shadow effect for the input | `boolean`                                | `false`                                            |
| variant  | Style variant of the input         | `'filled' \| 'outlined' \| 'borderless'` | Dark mode: `'filled'`<br/>Light mode: `'outlined'` |

InputNumber inherits all properties from Ant Design's InputNumber component.

### Input.Password

| Property | Description                        | Type                                     | Default                                            |
| -------- | ---------------------------------- | ---------------------------------------- | -------------------------------------------------- |
| shadow   | Enable shadow effect for the input | `boolean`                                | `false`                                            |
| variant  | Style variant of the input         | `'filled' \| 'outlined' \| 'borderless'` | Dark mode: `'filled'`<br/>Light mode: `'outlined'` |

InputPassword inherits all properties from Ant Design's Input.Password component.

### Input.OPT

| Property | Description                        | Type                                     | Default                                            |
| -------- | ---------------------------------- | ---------------------------------------- | -------------------------------------------------- |
| shadow   | Enable shadow effect for the input | `boolean`                                | `false`                                            |
| variant  | Style variant of the input         | `'filled' \| 'outlined' \| 'borderless'` | Dark mode: `'filled'`<br/>Light mode: `'outlined'` |

InputOPT inherits all properties from Ant Design's Input.OTP component.
