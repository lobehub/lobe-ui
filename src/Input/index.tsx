export { default } from './Input';
export { default as InputNumber } from './InputNumber';
export { default as InputOPT } from './InputOPT';
export { default as InputPassword } from './InputPassword';
export { default as TextArea } from './TextArea';
export type * from './type';
