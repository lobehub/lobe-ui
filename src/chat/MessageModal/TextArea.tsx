import { memo, useState } from 'react';

import { TextArea, TextAreaProps } from '@/Input';

interface MessageTextAreaProps extends Omit<TextAreaProps, 'onChange' | 'onBlur' | 'value'> {
  defaultValue: string;
  onChange?: (value: string) => void;
}
const MessageTextArea = memo<MessageTextAreaProps>(({ defaultValue, onChange, ...rest }) => {
  const [text, setText] = useState<string>(defaultValue || '');

  return (
    <TextArea
      onBlur={() => onChange?.(text)}
      onChange={(e) => {
        setText(e.target.value);
      }}
      onPressEnter={() => {
        onChange?.(text);
      }}
      {...rest}
      value={text}
    />
  );
});

export default MessageTextArea;
