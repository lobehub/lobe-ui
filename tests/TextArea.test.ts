import { render } from '@testing-library/react';
import React from 'react';
import { describe, expect, it } from 'vitest';

import { TextArea } from '../src/Input/TextArea';

describe('TextArea', () => {
  it('should render correctly with default props', () => {
    const { container } = render(
      React.createElement(TextArea, { 'data-testid': 'textarea' } as any),
    );
    expect(container.firstChild).toMatchInlineSnapshot(`
      <textarea
        class="ant-input css-dev-only-do-not-override-1v5z42l ant-input-outlined acss-0"
        data-testid="textarea"
        style="resize: none;"
      />
    `);
  });

  it('should render with resize enabled', () => {
    const { container } = render(
      React.createElement(TextArea, { 'data-testid': 'textarea', 'resize': true } as any),
    );
    const style = window.getComputedStyle(container.firstChild as HTMLElement);
    // When resize is true, the style.resize should not be 'none'
    expect(style.resize).not.toBe('none');
  });

  it('should render with resize disabled', () => {
    const { container } = render(
      React.createElement(TextArea, { 'data-testid': 'textarea', 'resize': false } as any),
    );
    const style = window.getComputedStyle(container.firstChild as HTMLElement);
    expect(style.resize).toBe('none');
  });

  it('should render with custom className', () => {
    const { container } = render(
      React.createElement(TextArea, {
        'className': 'custom-class',
        'data-testid': 'textarea',
      } as any),
    );
    expect((container.firstChild as HTMLElement).className).toMatch(/custom-class/);
  });

  it('should render with custom style', () => {
    const { container } = render(
      React.createElement(TextArea, {
        'data-testid': 'textarea',
        'style': { color: 'red' },
      } as any),
    );
    const style = window.getComputedStyle(container.firstChild as HTMLElement);
    // computed color might be rgb, so check for substring
    expect(style.color.replaceAll(' ', '')).toMatch(/red|rgb\(255,0,0\)/i);
  });

  it('should render with shadow', () => {
    const { container } = render(
      React.createElement(TextArea, { 'data-testid': 'textarea', 'shadow': true } as any),
    );
    expect(container.firstChild).toMatchInlineSnapshot(`
      <textarea
        class="ant-input css-dev-only-do-not-override-1v5z42l ant-input-outlined acss-0"
        data-testid="textarea"
        style="resize: none;"
      />
    `);
  });

  it('should render with different variants', () => {
    const variants = ['filled', 'outlined', 'borderless', 'underlined'] as const;
    const snapshots: string[] = [];
    for (const variant of variants) {
      const { container } = render(
        React.createElement(TextArea, { 'data-testid': 'textarea', variant } as any),
      );
      // Collect the rough snapshot for each variant for inline snapshot assertion
      snapshots.push((container.firstChild as HTMLElement).outerHTML);
    }
    // All variants for inline snapshot
    expect(snapshots).toMatchInlineSnapshot([
      `<textarea data-testid="textarea" style="resize: none;" class="ant-input css-dev-only-do-not-override-1v5z42l ant-input-filled acss-waauh5"></textarea>`,
      `<textarea data-testid="textarea" style="resize: none;" class="ant-input css-dev-only-do-not-override-1v5z42l ant-input-outlined acss-0"></textarea>`,
      `<textarea data-testid="textarea" style="resize: none;" class="ant-input css-dev-only-do-not-override-1v5z42l ant-input-borderless acss-v7c7v6"></textarea>`,
      `<textarea data-testid="textarea" style="resize: none;" class="ant-input css-dev-only-do-not-override-1v5z42l ant-input-underlined acss-0"></textarea>`,
    ]);
  });

  it('should merge custom style with resize style', () => {
    const { container } = render(
      React.createElement(TextArea, {
        'data-testid': 'textarea',
        'resize': false,
        'style': { color: 'blue' },
      } as any),
    );
    const style = window.getComputedStyle(container.firstChild as HTMLElement);
    expect(style.resize).toBe('none');
    // computed color might be rgb, so check for substring
    expect(style.color.replaceAll(' ', '')).toMatch(/blue|rgb\(0,0,255\)/i);
  });
});
