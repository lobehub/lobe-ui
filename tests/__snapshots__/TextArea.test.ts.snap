// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TextArea > should render correctly with default props 1`] = `
<textarea
  class="ant-input css-dev-only-do-not-override-1v5z42l ant-input-outlined acss-0"
  style="resize: none;"
/>
`;

exports[`TextArea > should render with different variants 1`] = `
<textarea
  class="ant-input css-dev-only-do-not-override-1v5z42l ant-input-filled acss-waauh5"
  style="resize: none;"
/>
`;

exports[`TextArea > should render with different variants 2`] = `
<textarea
  class="ant-input css-dev-only-do-not-override-1v5z42l ant-input-outlined acss-0"
  style="resize: none;"
/>
`;

exports[`TextArea > should render with different variants 3`] = `
<textarea
  class="ant-input css-dev-only-do-not-override-1v5z42l ant-input-borderless acss-v7c7v6"
  style="resize: none;"
/>
`;

exports[`TextArea > should render with different variants 4`] = `
<textarea
  class="ant-input css-dev-only-do-not-override-1v5z42l ant-input-underlined acss-0"
  style="resize: none;"
/>
`;

exports[`TextArea > should render with shadow 1`] = `
<textarea
  class="ant-input css-dev-only-do-not-override-1v5z42l ant-input-outlined acss-0"
  style="resize: none;"
/>
`;
